google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreImpl.java

```diff
@@ -126,11 +138,24 @@ public DocumentReference document(@Nonnull String documentPath) {
 
     ApiStreamObserver<BatchGetDocumentsResponse> responseObserver =
         new ApiStreamObserver<BatchGetDocumentsResponse>() {
+          int numResponses;
```

Comment:
Is onNext guaranteed to run on one thread for all responses? If not you probably need to make this atomic to make sure the increment is correct + the latest value is visible to other threads.